openmc/tallies.py

```diff
@@ -0,0 +1,3204 @@
+            self.sum
+
+        if self.sparse:
+            return np.reshape(self._sum_sq.toarray(), self.shape)
+        else:
+            return self._sum_sq
+
+    @property
+    def mean(self):
+        if self._mean is None:
+            if not self._sp_filename:
+                return None
+
+            self._mean = self.sum / self.num_realizations
+
+            # Convert NumPy array to SciPy sparse LIL matrix
+            if self.sparse:
+                self._mean = sps.lil_matrix(self._mean.flatten(),
+                                            self._mean.shape)
+
+        if self.sparse:
+            return np.reshape(self._mean.toarray(), self.shape)
+        else:
+            return self._mean
+
+    @property
+    def std_dev(self):
+        if self._std_dev is None:
+            if not self._sp_filename:
+                return None
+
+            n = self.num_realizations
+            nonzero = np.abs(self.mean) > 0
+            self._std_dev = np.zeros_like(self.mean)
+            self._std_dev[nonzero] = np.sqrt((self.sum_sq[nonzero]/n -
+                                              self.mean[nonzero]**2)/(n - 1))
+
+            # Convert NumPy array to SciPy sparse LIL matrix
+            if self.sparse:
+                self._std_dev = sps.lil_matrix(self._std_dev.flatten(),
+                                               self._std_dev.shape)
+
+            self.with_batch_statistics = True
+
+        if self.sparse:
+            return np.reshape(self._std_dev.toarray(), self.shape)
+        else:
+            return self._std_dev
+
+    @property
+    def with_batch_statistics(self):
+        return self._with_batch_statistics
+
+    @property
+    def derived(self):
+        return self._derived
+
+    @property
+    def derivative(self):
+        return self._derivative
+
+    @property
+    def sparse(self):
+        return self._sparse
+
+    @estimator.setter
+    def estimator(self, estimator):
+        cv.check_value('estimator', estimator, ESTIMATOR_TYPES)
+        self._estimator = estimator
+
+    @triggers.setter
+    def triggers(self, triggers):
+        cv.check_type('tally triggers', triggers, MutableSequence)
+        self._triggers = cv.CheckedList(openmc.Trigger, 'tally triggers',
+                                        triggers)
+
+    @name.setter
+    def name(self, name):
+        if name is not None:
+            cv.check_type('tally name', name, str)
+            self._name = name
+        else:
+            self._name = ''
+
+    @derivative.setter
+    def derivative(self, deriv):
+        if deriv is not None:
+            cv.check_type('tally derivative', deriv, openmc.TallyDerivative)
+        self._derivative = deriv
+
+    @filters.setter
+    def filters(self, filters):
+        cv.check_type('tally filters', filters, MutableSequence)
+
+        # If the filter is already in the Tally, raise an error
+        for i, f in enumerate(filters[:-1]):
```

Comment:
You could have a separate type checker for ensuring a `CheckedList` has no duplicates. This would remove the need to implement this same logic over and over again in this file. Also, a more efficient way to check for duplicates would be to create a `set` out of filters and check if each element is already in the set:

```python
visited_filters = set()
for f in filters:
    if f in visited_filters:
        # raise error
    visited_filters.add(f)
```

This implementation is `O(n)` rather than `O(n^2)`.

** Ditto for `nuclides`, `scores`, and wherever else this functionality is desired.